sw/device/tests/sim_dv/uart_tx_rx_test.c

```diff
@@ -0,0 +1,350 @@
+                          bytes_remaining, &bytes_written_isr);
+  CHECK(status, "dif_uart_bytes_send failed");
+  uart_tx_bytes_written += bytes_written_isr;
+}
+
+/**
+ * Receive data in over RX without blocking.
+ *
+ * Non-blocking function that reads bytes received over UART RX port. It reads
+ * as many bytes as there are available in the RX FIFO and returns the number of
+ * bytes read, which is used the next time this function is called. Eventually,
+ * all UART_DATASET_SIZE bytes need to be read.
+ */
+void receive_rx_data_nonblocking(void) {
+  size_t bytes_remaining = sizeof(exp_uart_rx_data) - uart_rx_bytes_read;
+  // Return if we have no data to send.
+  if (bytes_remaining == 0)
+    return;
+  // Receive the remaining uart_rx_data.
+  size_t bytes_read_isr = 0;
+  bool status = dif_uart_bytes_receive(&uart0, bytes_remaining,
+                                       &uart_rx_data[uart_rx_bytes_read],
+                                       &bytes_read_isr);
+  CHECK(status, "dif_uart_bytes_receive failed");
+  uart_rx_bytes_read += bytes_read_isr;
+}
+
+/**
+ * Provides external irq handling for this test.
+ *
+ * This function overrides the default external irq handler in
+ * `sw/device/lib/handler.h`.
+ */
+void handler_irq_external(void) {
+  // Find which interrupt fired at PLIC by claiming it.
+  dif_plic_irq_id_t plic_irq_id;
+  bool status =
+      dif_plic_irq_claim(&plic0, kTopEarlgreyPlicTargetIbex0, &plic_irq_id);
+  CHECK(status, "dif_plic_irq_claim failed");
+
+  // Check if it is the right peripheral.
+  top_earlgrey_plic_peripheral_t peripheral = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  CHECK(peripheral == kTopEarlgreyPlicPeripheralUart,
+        "Only UART is expected to generate interrupts in this test!");
+
+  // Correlate the interrupt fired at PLIC with UART.
+  dif_uart_interrupt_t uart_irq = (dif_uart_interrupt_t)(
+      (int)plic_irq_id - (int)kTopEarlgreyPlicIrqIdUartTxWatermark);
+  switch (uart_irq) {
+    case kDifUartInterruptTxWatermark:
+      uart_irq_tx_watermark_fired = true;
+      break;
+    case kDifUartInterruptRxWatermark:
+      uart_irq_rx_watermark_fired = true;
+      break;
+    case kDifUartInterruptTxEmpty:
+      uart_irq_tx_empty_fired = true;
+      break;
+    case kDifUartInterruptRxOverflow:
+      uart_irq_rx_overflow_fired = true;
+      break;
+    default:
+      LOG_ERROR("Unsupported unterrupt fired");
+      test_status_set(kTestStatusFailed);
+  }
+
+  // Check if the same interrupt fired at UART as well.
+  dif_uart_enable_t irq_state;
+  dif_uart_irq_state_get(&uart0, uart_irq, &irq_state);
+  CHECK(irq_state == kDifUartEnable,
+        "UART interrupt at fired at PLIC did not fire at UART");
+
+  // Clear the interrupt at UART.
+  dif_uart_irq_state_clear(&uart0, uart_irq);
+
+  // Complete the IRQ at PLIC.
+  dif_plic_irq_complete(&plic0, kTopEarlgreyPlicTargetIbex0, &plic_irq_id);
```

Comment:
I think with recent change `plic_irq_id` can be passed as a value not a pointer. It used to be a structure, now it's simple `uint32_t`. This requires a change to the corresponding plic function. Probably could just be a separate commit in this change, or should it be a separate PR?

I can do it, or you could do it if you think it would be easier this way. I am happy either way, but please let me know.